mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -346,9 +347,206 @@ propagateUnmergeVectorization(T &&dimAndLength,
+        auto &&zip = llvm::zip(params, upperDims);
+        SmallVector<std::tuple<int64_t, uint32_t>> data(zip.begin(), zip.end());
+        std::sort(data.begin(), data.end());
+        bool maybeUnmerge = true;
+
+        // Verify that the Embed is an Unmerge operation and
+        // at the same time create the sorted unmerge params
+        if (!std::all_of(params.begin(), params.end(),
+                         [](auto p) { return p > 0; })) {
+          break;
+        }
+
+        SmallVector<int64_t> unmergeParams;
+        SmallVector<uint32_t> unmergeDims;
+        // Embed parameters should be in the form of
+        // [1, l_0, l_0*l_1, ...] to be considered a valid
+        // unmerge
+        for (size_t i = 0; i < data.size() - 1; i++) {
+          auto paramI = std::get<0>(data[i]);
+          auto paramI1 = std::get<0>(data[i + 1]);
+          auto dim = std::get<1>(data[i]);
+          auto unmergeParam = paramI1 / paramI;
+          if (unmergeParam * paramI != paramI1) {
+            maybeUnmerge = false;
+            break;
+          }
+          unmergeParams.push_back(unmergeParam);
+          unmergeDims.push_back(dim);
+        }
+
+        // We are not done yet. To be sure this is an unmerge
+        // the faster parameter needs to be 1 and we should set
+        // the upper bound
+        auto fasterParam = std::get<0>(data[0]);
+        if (maybeUnmerge && fasterParam == 1 && upperBounds.size()) {
+          auto slowerDim = std::get<1>(data.back());
+          unmergeParams.push_back(upperBounds[0]);
+          unmergeDims.push_back(slowerDim);
+
+          // We have the unmerge dimensions/params going from the faster
+          // to the slower, but they should be in the exact reverse order
+          std::reverse(unmergeParams.begin(), unmergeParams.end());
+          std::reverse(unmergeDims.begin(), unmergeDims.end());
+
+          // Now we can use the parameters as unmerge parameters and see if
+          // there are contiguous dimensions to be folded together.
+          findCountiguousGroupsUnmerge(unmergeDims, unmergeParams, dimToMerge,
+                                       contiguousGroups);
+        }
+        break;
+      }
+
+      case TransformType::PassThrough:
+      case TransformType::Pad:
+      case TransformType::Broadcast:
+      case TransformType::Slice:
+        // We only care about how these transformations shuffle
+        // the dimensions
+        for (auto pair : llvm::zip(upperDims, lowerDims)) {
+          auto u = std::get<0>(pair);
+          auto l = std::get<1>(pair);
+          thisDimToMerge[l] = dimToMerge[u];
+        }
+        break;
+        break;
+      case TransformType::Unmerge:
+        findCountiguousGroupsUnmerge(upperDims, params, dimToMerge,
+                                     contiguousGroups);
+        break;
+      }
+    }
+    dimToMerge = thisDimToMerge;
+  }
+
+  // Last global unmerge
+  SmallVector<uint32_t> sortedDims;
```

Comment:
Couldn't this just be `[0, 1, 2, (outputShape.size() - 1]`?

Reply:
Unfortunately not (I think), because of the AddDim. If we unmerge{0,1,2,3} but we previously had AddDim [1] it will fail.